executor/insert_common.go

```diff
@@ -619,41 +620,20 @@ func (e *InsertValues) isAutoNull(ctx context.Context, d types.Datum, col *table
 	return false
 }
 
-func (e *InsertValues) hasAutoIncrementColumn() (int, bool) {
-	colIdx := -1
-	for i, c := range e.Table.Cols() {
+func findAutoIncrementColumn(t table.Table) (col *table.Column, offsetInRow int, found bool) {
+	for i, c := range t.Cols() {
 		if mysql.HasAutoIncrementFlag(c.Flag) {
-			colIdx = i
-			break
+			return c, i, true
 		}
 	}
-	return colIdx, colIdx != -1
+	return nil, -1, false
 }
 
-func (e *InsertValues) lazyAdjustAutoIncrementDatumInRetry(ctx context.Context, rows [][]types.Datum, colIdx int) ([][]types.Datum, error) {
-	// Get the autoIncrement column.
-	col := e.Table.Cols()[colIdx]
-	// Consider the colIdx of autoIncrement in row are the same.
-	length := len(rows)
-	for i := 0; i < length; i++ {
-		autoDatum := rows[i][colIdx]
-
-		// autoID can be found in RetryInfo.
-		retryInfo := e.ctx.GetSessionVars().RetryInfo
-		if retryInfo.Retrying {
-			id, err := retryInfo.GetCurrAutoIncrementID()
-			if err != nil {
-				return nil, err
-			}
-			autoDatum.SetAutoID(id, col.Flag)
-
-			if err = col.HandleBadNull(&autoDatum, e.ctx.GetSessionVars().StmtCtx); err != nil {
-				return nil, err
-			}
-			rows[i][colIdx] = autoDatum
-		}
-	}
-	return rows, nil
+func setDatumAutoIDWithCast(ctx sessionctx.Context, d *types.Datum, id int64, col *table.Column) error {
```

Comment:
Do we need `HandleBadNull`?

Reply:
Since we `d.SetAutoID` explicitly , the datum must be not null. `HandleBadNull` is meaningless.